### ISSUE TITLE
1:1 Map only maps to vJoy device 1.

### ISSUE THREAD

**Issue by Zarklord** [WhiteMagic/JoystickGremlin#238]
It should either ask what vJoy device to map to, or somehow know what vJoy device to map it to.

**Comment by WhiteMagic** [comment]
Yeah the 1:1 mapping doesn't look for what vJoy devices exist, it likely also fail if vJoy device 1 does not exist. Another user actually modified Gremlin do have a selection UI to pick which vJoy device to use for the 1:1 mapping but I haven't gotten a pull request from him. However, with the likely changes the UI portion of that change would be scrapped, though the code logic would ideally remain.

**Comment by code-monet** [comment]
I can take a stab at implementing this in R14 if you like - it seems the code wasn't copied over from R13. I should be able to do the Python parts but might need help with the QMLs.

I was thinking:

1. It should be a menu item, and open a dialog box.
2. Dialog box should allow picking the input device to map from, and the vJoy ID to map to
3. [maybe?] Allow specifying ranges of buttons (because sometimes devices list buttons and axes in Windows that don't exist, or only exist if specific accessories are connected).
4. [maybe?] A checkbox for whether to replace existing mappings from the source device (default unchecked).
5. [maybe?] A checkbox for whether to overwrite mappings to the target vJoy device to maintain 1:1 (default unchecked).

**Comment by WhiteMagic** [comment]
You're more than welcome but be prepared to fight QML quite a bit, though the one for this should be not too bad.

1. Yes, an entry under tools that opens a dialog would be the correct approach, as the tool needs to be a bit smarter than it used to be.
2. Picking input and output devices is a bit trickier though I think there is roughly two main usage modes
   a) One physical device to one vJoy device
   b) Multiple physical devices sequentially onto 1 (or more) vJoy devices
So far Gremlin was doing a simplified version of b), mapping all physical devices onto vJoy 1, repeatedly using the first input if it ran out of vJoy inputs. This behavior is probably the least desirable.

One possible flow I could imagine is this, assuming a UI that looks something like the one below (just a quick idea but by no means a guide of what it has to be like):
1. Pick Combination of physical and virtual devices (have to have at least one of each)
2. Process each input on the physical device, adding a `Map to vJoy` instance based on the toggles (overwrite / repeat)
3. Stop when no more valid vJoy targets exist (when not repeating)
4. Stop when no more physical inputs remain

This should be able to handle use case 1 naturally by picking combinations one after the other and pressing the "Add Mappings" button in between. Use case 2 would be achieve by just picking the desire physical and virtual devices and the logic should do one after the other, switching to using vJoy device 2, 3, etc. when the previous vJoy devices "run out of" inputs to use. The UI as drawn does not support picking an order and it's implementation dependant but a user could simply keep the vJoy devices fixed and select one physical input after the other to get the ordering they desire. One could get fancy here with drop downs or drag&drop reordering, but I would say that's fancier than needed, at least for a v1.

<img width="609" height="433" alt="Image" src="https://github.com/user-attachments/assets/c08f137c-f22a-420f-89f7-c3f95406d867" />

On the logic side you'll need code that is capable of figuring out which vJoy devices are used in profiles. The old code for that is of no use anymore as one needs to look at the profile and that changed completely. You'll need to find all `map_to_vjoy` actions and then check which of these are actively used by physical inputs, as actions may exist without being used. Then you obviously need to find the one that matches the desired input type and also handle the case of no input existing. Creating the actual action and tying it to the input shouldn't be too hard for you since you've played with those parts with the unit tests already.

For the QML bit `DialogCalibration.qml` and `DialogManageModes.qml` are probably somewhat decent starting points, though the calibration one may have broken again. All these files have been written with quite some time apart and I've figured out more things over time so not everything they do is good or how I'd do it now. But that's something we can iterate, also to make life easy, I'd suggest adding a button to the toolbar for yourself to open the window quickly when launching Gremlin.

**Comment by code-monet** [comment]
Thanks for the detailed feedback; I'll give this a stab over the next few days. A couple of points of clarification regarding your "logic" paragraph:

1. In addition to`map_to_vjoy`, I will also want to check `macro` actions that don't use `map_to_vjoy` but do actually "map to vjoy".
2. Do we want to ignore bindings in loaded user scripts, or take them into account?
3. When you say "handle the case of no input existing" - do you mean that vJoy doesn't have any inputs of that type configured (e.g. no hats), or that the action has no active binding (or the physical input is no longer connected)? If the profile has a binding but the device is disconnected, do we want to ignore those vJoy inputs, or bring them back into the "available" pool?

**Comment by WhiteMagic** [comment]
The overarching thing to keep in mind is that this feature's primary purpose is to get people up and running when they don't have complicated setups. So users with involved setups should not expect that tool magically maps everything just right. It should get you up and running quicker if all you do is simple remaps, but beyond that, I would not want to impose on the tool to do "the right thing", mainly because at that point "the right thing" is up to interpretation.
 
1. You can but then the question becomes at what point do you consider it being used. Like is a complex macro using it or only simple single press/release ones? I'd also be fine with ignoring macros entirely.
2. Yes, I'd ignore them. This could be a feature addition, or possibly more useful the option to "blacklist"/"whitelist" vJoy inputs to be used.
3. Oof that's a few cases
   a. I was thinking of "no hats on vJoy but physical has a hat" mainly.
   b. I would consider vJoy bindings on disconnected devices as being not in use, thus fair game to be assigned. Gremlin never clears existing bindings and with the new system there is even more ways to have "dead"/unused assignments. So I'd only consider the case of connected devices, again because of the main use case is simple setups and getting people up and running.
   c. An action that's not bound to an input should be reused in my mind.

**Comment by code-monet** [comment]
What should be the handling for multiple modes? Almost certainly the bindings should only be created for the current mode; and I'm thinking we should just ignore the other modes completely i.e. physical and vJoy inputs used in other modes should be assumed available for this operation.

Also, thinking of calling it `auto_map`, since it does more than 1:1 now. Let's reserve `quick_map` for a different feature if we ever want to implement it (Gremlin cycles through vJoy inputs, user presses the input they want assigned).

**Comment by WhiteMagic** [comment]
Hmm, I don't even remember what previous versions of Gremlin did in that regard. There may be an argument for letting the user select the mode from a dropdown, but that then causes possible expectations of how inputs are reused. On the other hand it using the currently active mode may also be a bit confusing. I think having a text field stating what will happen, i.e. "Creating map to vjoy actions in all empty inputs to the ${current mode} mode" or similar. It informs but doesn't give direct control, might be the best compromise.

I think ignoring other modes for availability resolution is fine when we look at this tool as the "get Gremlin to be useful quickly for a newcomer". Such a user shouldn't have crazy mode setups etc. and someone that does likely will be able to correct and understand what's going on.

The name is fine with me, we could also use something like pass through in it as that's effectively what we're doing but maybe that's a bit more convoluted of a name than auto_map.

**Comment by code-monet** [comment]
([Still working on this](https://github.com/WhiteMagic/JoystickGremlin/compare/develop...code-monet:JoystickGremlin:auto_mapper); I have the Python code mostly down, and will start work on the UI next).

---

Oct 6: Basic UI is in place, need to add some user notifications and look over for any obvious bugs.
